File: LoneHPManagementBackend/src/main/java/com/prj/LoneHPManagement/Service/CollateralService.java
package com.prj.LoneHPManagement.Service;

import com.prj.LoneHPManagement.model.dto.CollateralRequest;
import com.prj.LoneHPManagement.model.dto.CollateralUpdateRequest;
import com.prj.LoneHPManagement.model.entity.Collateral;
import org.springframework.data.domain.Page;

import java.util.List;

public interface CollateralService {
    Collateral updateCollateral(int id, CollateralUpdateRequest request);
    //     Collateral saveCollateral(Collateral collateral);
 Collateral create(CollateralRequest dto);
   Page<Collateral> getCollateralsByLoanId(int loanId, int page, int size, String sortBy);
    List<Collateral> getAllCollaterals();
    Page<Collateral> getCollateralsByCifId(int cifId, int page, int size, String sortBy);
    Collateral getCollateralById(int id);
    List<Collateral> getCollateralByCIFId(int cifid);
    List<Collateral> getCollateralByLoanId(int loanId);

    void deleteCollateral(int id);
}

File: LoaneHpManagementFrontEnd/src/app/services/collateral.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Collateral } from '../models/collateral.model';
import { map } from 'rxjs/operators';
import { ApiResponse, PagedResponse } from '../models/common.types';

@Injectable({
  providedIn: 'root'
})
export class CollateralService {
  private apiUrl = `${environment.apiUrl}/collaterals`;

  constructor(private http: HttpClient) {}

  getCollateralsByLoanId(
    loanId: number, 
    page: number = 0, 
    size: number = 10, 
    sortBy: string = 'id'
  ): Observable<PagedResponse<Collateral>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy);

    return this.http.get<ApiResponse<PagedResponse<Collateral>>>(`${this.apiUrl}/loan/${loanId}`, { params })
      .pipe(
        map(response => response.data)
      );
  }

  getCollateralsByCifId(
    cifId: number, 
    page: number = 0, 
    size: number = 10, 
    sortBy: string = 'id'
  ): Observable<PagedResponse<Collateral>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy);

    return this.http.get<ApiResponse<PagedResponse<Collateral>>>(`${this.apiUrl}/cif/${cifId}`, { params })
      .pipe(
        map(response => response.data)
      );
  }
  getCollateralsByCifIdToSelect(cifId: number): Observable<Collateral[]> {
    return this.http.get<ApiResponse<Collateral[]>>(`${this.apiUrl}/cif/${cifId}/toSelect`)
      .pipe(
        map(response => response.data)
      );
  }

  createCollateral(collateral: Partial<Collateral>): Observable<Collateral> {
    return this.http.post<ApiResponse<Collateral>>(`${this.apiUrl}/create`, collateral)
      .pipe(
        map(response => response.data)
      );
  }

  updateCollateral(id: number, collateral: Partial<Collateral>): Observable<Collateral> {
    return this.http.put<ApiResponse<Collateral>>(`${this.apiUrl}/update/${id}`, collateral)
      .pipe(
        map(response => response.data)
      );
  }

  deleteCollateral(id: number): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`)
      .pipe(
        map(response => {
          if (response.status !== 200) {
            throw new Error(response.message || 'Error deleting collateral');
          }
        })
      );
  }
} 